### PR TITLE
Scottx611x/tophat2 edit

### PR DESCRIPTION
**Pull request by scottx611x**
This PR edits the Tophat2 tool-wrapper/macro to use gene annotation data. The file(s) are fetched via a data_manager that is introduced in this PR.

### CODE REVIEW

File: tools/tophat2/tophat_macros.xml [hunk 1]
```diff
@@ -42,7 +42,9 @@
...
+            <param format="gtf,gff3" name="gene_annotation_model" type="select" label="Gene Model Annotations" help="-G/--GTF; TopHat with a set of gene model annotations and/or known transcripts, as a GTF 2.2 or GFF3 formatted file. If this option is provided, TopHat will first extract the transcript sequences and use Bowtie to align reads to this virtual transcriptome first. Only the reads that do not fully map to the transcriptome will then be mapped on the genome. The reads that did map on the transcriptome will be converted to genomic mappings (spliced as needed) and merged with the novel mappings and junctions in the final tophat output. Please note that the values in the first column of the provided GTF/GFF file (column which indicates the chromosome or contig on which the feature is located), must match the name of the reference sequence in the Bowtie index you are using with TopHat.">
```
Comment (bgruening) [inline tools/tophat2/tophat_macros.xml]: @scottx611x can we make this optional so that every user can also use a plain GTF file.

**Comment by bgruening** [comment]
Great addition! Thanks @scottx611x!
@blankenberg can you please review it?

File: data_managers/data_manager_fetch_tophat2_gene_annotation/shed.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+categories:
```
Comment (nsoranzo) [inline data_managers/data_manager_fetch_tophat2_gene_annotation/shed.yml]: This file should called `.shed.yml` with an initial dot.

File: tools/tophat2/tophat_macros.xml (hunk 1 above)
Comment (nsoranzo) [inline tools/tophat2/tophat_macros.xml]: Ideally you can modify the `gene_model_ann` conditional giving 3 options: `No`, `Use a built-in gene annotation`, `Use a gene annotation from history`.

File: tools/tophat2/tophat_macros.xml [hunk 3]
```diff
@@ -38,12 +38,18 @@
...
+                <options from_data_table="tophat2_gene_annotation"/>
+             </param>
           </when>
+          <when value="Use a gene annotation from history">
+            <param name="ownFile" type="data" format="gtf,gff3" metadata_name="dbkey" label="Select the gene annotation from history" />
```
Comment (nsoranzo) [inline tools/tophat2/tophat_macros.xml]: For backwards compatibility you should have `name="gene_annotation_model"` .

File: tools/tophat2/tophat_macros.xml [hunk 4]
```diff
@@ -38,12 +38,18 @@
...
+          <when value="Use a built-in gene annotation">
+            <param format="gtf,gff3" name="gene_annotation_model" type="select" label="Gene Model Annotations" help="-G/--GTF; TopHat with a set of gene model annotations and/or known transcripts, as a GTF 2.2 or GFF3 formatted file. If this option is provided, TopHat will first extract the transcript sequences and use Bowtie to align reads to this virtual transcriptome first. Only the reads that do not fully map to the transcriptome will then be mapped on the genome. The reads that did map on the transcriptome will be converted to genomic mappings (spliced as needed) and merged with the novel mappings and junctions in the final tophat output. Please note that the values in the first column of the provided GTF/GFF file (column which indicates the chromosome or contig on which the feature is located), must match the name of the reference sequence in the Bowtie index you are using with TopHat.">
```
Comment (nsoranzo) [inline tools/tophat2/tophat_macros.xml]: I'd move the `help="-G--GTF; ..."` to `<param name="use_annotations"...>` above.

File: tools/tophat2/tophat_macros.xml [hunk 5]
```diff
@@ -36,14 +36,20 @@
...
+          <param name="use_annotations" type="select" label="Use Gene Annotation Model" help="-G/--GTF; TopHat with a set of gene model annotations and/or known transcripts, as a GTF 2.2 or GFF3 formatted file. If this option is provided, TopHat will first extract the transcript sequences and use Bowtie to align reads to this virtual transcriptome first. Only the reads that do not fully map to the transcriptome will then be mapped on the genome. The reads that did map on the transcriptome will be converted to genomic mappings (spliced as needed) and merged with the novel mappings and junctions in the final tophat output. Please note that the values in the first column of the provided GTF/GFF file (column which indicates the chromosome or contig on which the feature is located), must match the name of the reference sequence in the Bowtie index you are using with TopHat.">
             <option value="No">No</option>
-            <option value="Yes">Yes</option>
+            <option value="Use a built-in gene annotation">Use a built-in gene annotation</option>
+            <option value="Use a gene annotation from history">Use a gene annotation from history</option>
```
Comment (nsoranzo) [inline tools/tophat2/tophat_macros.xml]: Having option values with spaces and this long is probably not good practice, I'd use something like:
```
            <option value="indexed">Use a built-in gene annotation</option>
            <option value="history">Use a gene annotation from history</option>
```

File: tools/tophat2/tophat2_wrapper.xml [hunk 6]
```diff
@@ -61,7 +61,7 @@
             ## Supplying junctions parameters.
             #if $params.own_junctions.use_junctions == "Yes":
                 #if $params.own_junctions.gene_model_ann.use_annotations == "Yes":
-                    -G $params.own_junctions.gene_model_ann.gene_annotation_model
+                    -G $params.own_junctions.gene_model_ann.gene_annotation_model.fields.path
                 #end if
```
Comment (nsoranzo) [inline tools/tophat2/tophat2_wrapper.xml]: This code should be updated to the 2 new option values:
```
                #if $params.own_junctions.gene_model_ann.use_annotations == "indexed"
                    -G $params.own_junctions.gene_model_ann.gene_annotation_model.fields.path
                #elif $params.own_junctions.gene_model_ann.use_annotations == "history"
                    -G $params.own_junctions.gene_model_ann.gene_annotation_model
                #end if
```

File: tools/tophat2/tophat_macros.xml [hunk 7]
```diff
@@ -36,14 +36,20 @@
...
+          <when value="indexed">
+            <param format="gtf,gff3" name="gene_annotation_model" type="select" label="Gene Model Annotations">
+                <options from_data_table="tophat2_gene_annotation"/>
+             </param>
           </when>
+          <when value="history">
+            <param name="gene_annotation_model" type="data" format="gtf,gff3" metadata_name="dbkey" label="Select the gene annotation from history" />
```
Comment (nsoranzo) [inline tools/tophat2/tophat_macros.xml]: Can you also restore the label to "Gene Model Annotations", please?

**Comment by mvdbeek** [comment]
@scottx611x 
Would it be possible to rename the `tophat2_gene_annotation` data table to `gene_annotation` and the data manager to data_manager_fetch_gene_annotation ?
I think this would be useful for tools beyond tophat, like hisat2, featureCounts, htseq-counts and so on.
There is nothing really specific to tophat2 in this data manager, right?

**Comment by scottx611x** [comment]
@mvdbeek I made the edits you suggested earlier today, I appreciate your insight!

**Comment by scottx611x** [comment]
@mvdbeek No, there was nothing specific to tophat2 in within the data manager, but there is still the edited tophat2 tool-wrapper and macro in this PR

**Comment by scottx611x** [comment]
@mvdbeek @nsoranzo @bgruening Just curious as to the proceedings after this gets merged... Will my changes immediately be available via a toolshed (test/main)?

**Comment by nsoranzo** [comment]
@scottx611x No, someone from the [Galaxy Core Team](https://wiki.galaxyproject.org/GalaxyTeam) will have to push them to the ToolSheds.

**Comment by scottx611x** [comment]
My team was dependent on my changes being in a toolshed so I was asked to self publish.

Edited tophat2: https://toolshed.g2.bx.psu.edu/view/scottx611x/tophat2_with_gene_annotations/3d6395d31c96

Gene annotation data manager: https://toolshed.g2.bx.psu.edu/view/scottx611x/data_manager_fetch_gene_annotation/189164b8f9d2

**Comment by bgruening** [comment]
I'm fine with merging this. Anyone else? It violates our rules that we don't want to add new tools to the devteam repository but I guess we can make an exception as it is TopHat related.

File: data_managers/data_manager_fetch_gene_annotation/data_manager/data_manager.py [hunk 8]
```diff
@@ -16,13 +16,21 @@
...
 
 def url_download(url, workdir):
+    if not os.path.exists(workdir):
+        os.makedirs(workdir)
     file_path = os.path.join(workdir, 'download.dat')
     if not os.path.exists(workdir):
         os.makedirs(workdir)
     src = None
     dst = None
+    hdr = {'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11',
```
Comment (bgruening) [inline data_managers/data_manager_fetch_gene_annotation/data_manager/data_manager.py]: Out of curiosity why is this needed?

**Comment by bgruening** [comment]
ping @mvdbeek @blankenberg @nsoranzo

**Comment by nsoranzo** [comment]
:+1: for me. Should we wait for a review of the data manager by our official expert @blankenberg?

**Comment by bgruening** [comment]
I would like to wait but it seems @blankenberg is not following github issues involving tools anymore :(

**Comment by blankenberg** [comment]
I haven't tried running this, but just some quick notes.

The data table has dbkey defined, but it doesn't look like the data manager ever sets it.

Would be nice to mention that the "gene_annotations" are gff files, perhaps even as part of the table name.

Should probably use the `value_translation`s on the paths like other examples

Could also allow adding annotations from admin's history.

**Comment by bgruening** [comment]
@scottx611x do you plan to work on @blankenberg suggestions?
@blankenberg thanks Dan!

**Comment by scottx611x** [comment]
@bgruening Sorry, this got lost in all my notifications, I'll check it out tomorrow!

**Comment by nsoranzo** [comment]
@blankenberg Merge?

**Comment by bgruening** [comment]
@scottx611x sorry that we missed this so long. Thanks a bunch and happy new year!

**Comment by martenson** [comment]
@bgruening `data_manager_fetch_gene_annotation` is owned by @scottx611x so I am afraid I cannot

https://toolshed.g2.bx.psu.edu/view/scottx611x/data_manager_fetch_gene_annotation/16e961b2f35c

**Comment by scottx611x** [comment]
Hi all, I'm sorry if my development on this tool was not up to snuff. To be fair, I was a fairly new to the world of Galaxy when I wrote this. 

@martenson: I believe I can transfer ownership on the toolshed to you? Let me know if this is the best way to proceed.

**Comment by nsoranzo** [comment]
@scottx611x : Actually the `.shed.yml` file specifies `devteam` as the repo owner, so if you are OK with that, @martenson can create a new repo owned by `devteam` with `planemo shed_update --force_repository_creation` and then you can deprecate your repository using the ToolShed UI.

**Comment by martenson** [comment]
@nsoranzo the repo has `Times cloned / installed: 1253` - I think it would be a shame to disconnect/deprecate it

**Comment by martenson** [comment]
@nsoranzo I think we can give access to devteam/iuc via TS and adopt the repo in GitHub if it is requested.

**Comment by scottx611x** [comment]
I'm okay with whatever you all choose, I'm not going to have time to maintain or actively develop so I'll let you all make the call.

**Comment by scottx611x** [comment]
Also @nsoranzo,  a majority of the `Times cloned / installed` are most likely due to [refinery platform](https://github.com/refinery-platform/refinery-platform) deployments (which is mostly just in-house activity for now). Not sure if that affects your decision

**Comment by nsoranzo** [comment]
@scottx611x Can you grant authority to make changes to "devteam" user in the main repository page and add "devteam@galaxyproject.org" to the admin users? For the latter:
1) Go to the repo page, in the "Repository Actions" menu on the top right click on "Manager repository administrators"
2) In the page there is a textbox titled "Users associated with...", add "devteam@galaxyproject.org" from the box on its right.

**Comment by scottx611x** [comment]
@nsoranzo Done and Done

**Comment by nsoranzo** [comment]
Thanks @scottx611x! I've changed the owner of the repository in the `.shed.yml` file in commit https://github.com/galaxyproject/tools-devteam/commit/b1f84d7accfa17f88aed352469bfca81b48e4161 . Can you try again to shed_update @martenson?
